### PR TITLE
GH-1723: Allow for overriding the default dispatch choice

### LINKED ISSUE #1723: Allow different default operation for Fuseki endpoints
### Version

4.7.0

### Feature

Dispatching a request to an endpoint URL with multiple choices looks at the request to see which SPARQL operation (query, update, GSP) it is. The default is a plain HTTP request (REST-like) on the endpoint which is the quads extension of GSP (no `?graph=`, no `?default` - plain POST/GET etc with a content type that isn't registered (query and update are registered).

This minor enhancement is for Fuseki modules to provides for a different default operation for the plain REST-like operation. Usage includes endpoint specific error handling or extension code that can further decide the operation to execute.

Examples of multi-choice endpoints:
2 operations on the dataset URL `/dataset`:
```
    fuseki:endpoint [ fuseki:operation fuseki:query ; ] ;
    fuseki:endpoint [ fuseki:operation fuseki:update ; ] ;
```
or 2 on `/dataset/sparql`:
```
    fuseki:endpoint [ fuseki:operation fuseki:query ; fuseki:name "sparql" ] ;
    fuseki:endpoint [ fuseki:operation fuseki:update ; fuseki:name "sparql"; ] ;
```

### Are you interested in contributing a solution yourself?

Yes

### PR DESCRIPTION
**Pull request by afs**
GitHub issue resolved #1723

Pull request Description:
The PR allows a `DataService` to provide a custom handling of operations which are not recognized by the SPARQL protocols dispatcher.

Some other Fuseki items found during development are included.

----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx or JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

### CODE REVIEW

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/sys/FusekiModules.java [hunk 1]
```diff
@@ -38,35 +40,45 @@ public static void load() {
...
     /*package*/ static void forEachModule(Consumer<FusekiModule> action) {
-        if ( registry == null )
-            load();
-        if ( registry == null || registry.isEmpty() )
-            return ;
-        registry.forEach(action);
+        synchronized(lock) {
+            if ( registry == null )
+                load();
+            if ( registry == null || registry.isEmpty() )
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/sys/FusekiModules.java:79]: I think the `load()` call above means the `registry` cannot be null (it'll be at least an empty arraylist?)

**Review by kinow — APPROVED** [review]

File: jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java [hunk 2]
```diff
@@ -395,27 +455,13 @@ private static void applyAuthentication(String user, DataService dataService, En
...
+        return isReadMethod(request) ? GSP_R : GSP_RW;
     }
 
     /**
-     * Determine the {@link Operation} for a Quads operation. (GSP, except on the
-     * whole dataset).
-     * <p>
-     * Assumes, and does not check, that the action is a Quads action.
+     * Return whether request method is a "read" (GET or HEAD) or "write", modifying
+     * (POST, PUT, DELETE, PATCH)
```
[the remark this answers is not in the document]
Comment (afs) [reply jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java:463]: `ActionProcessor` has a switch on method type. TRACE isn't provided by any service IIRC. https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/TRACE questions whether it is used in browsers at all.

CONNECT doesn't get to the dispatcher.
(I'll add a definite rejection case to `ActionProcessor` to reject it rather than ignore it.)

It looks like OPTIONS works properly for everything except for multifunction endpoints for quads operations (this one case in `Dispatcher`).

We can't tell if it is "read" or "write" until the real operation is tried.

GSP_R vs GSP_RW is a weird case - one is a subset of the other.

So it gets "GSP_RW" which gives the full range of methods possible. Unfortunately, if only GSP_R is configured it will bounced because there is no implementation of GSP_RW.  Probably a corner case bug.

Having secure by-design read-only operation is more important than details of OPTIONS (which I never seen used on Fuseki - except by the test suite).

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/sys/FusekiModules.java (hunk 1 above)
Comment (afs) [reply jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/sys/FusekiModules.java:79]: Yes - but safe code in case `load` changes can leave the registry as null. Null tests are free! Its sort of left over from before `synchonized`. Multithreading startup complexity now avoided.

File: jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (afs) [reply jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java:463]: PS

Overloaded endpoint, GSP_RW: OPTIONS works.

Overloaded endpoint, GSP_R: OPTIONS is rejected with "405 HTTP method not allowed: OPTIONS" which isn't illegal, just unhelpful.

The dispatcher in just for SPARQL - it's a full HTTP REST environment! So I might fix it just because.

[the remark this answers is not in the document]
Comment (rvesse) [reply jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java:463]: `OPTIONS` being rejected is most likely to be a problem for browser based UIs since it's used as a pre-flight request by browsers prior to sending the real request

[the remark this answers is not in the document]
Comment (kinow) [reply jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java:463]: Thanks for the explanation & investigation @afs!

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/sys/FusekiModules.java (hunk 1 above)
Comment (kinow) [reply jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/sys/FusekiModules.java:79]: >Yes - but safe code in case load changes can leave the registry as null. Null tests are free! Its sort of left over from before synchonized. Multithreading startup complexity now avoided.

:+1: sounds good to me, thanks!

File: jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (afs) [reply jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java:463]: @rvesse -- CORS pre-flight is handled before Fuseki.

https://github.com/apache/jena/blob/ea9406a6ca5be4e34c6e7db57e282947be529450/jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/FusekiServer.java#L1452

It is only non-CORS OPTIONS here.

[the remark this answers is not in the document]
Comment (rvesse) [reply jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java:463]: > @rvesse -- CORS pre-flight is handled before Fuseki.
> 
> https://github.com/apache/jena/blob/ea9406a6ca5be4e34c6e7db57e282947be529450/jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/FusekiServer.java#L1452
> 
> It is only non-CORS OPTIONS here.

I suspected as much, thanks for confirming

File: jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java [hunk 3]
```diff
@@ -463,6 +509,8 @@ private static Operation gspOperation(HttpAction action, HttpServletRequest requ
      * (POST, PUT, DELETE, PATCH)
      */
     private static boolean isReadMethod(HttpServletRequest request) {
+        // OPTIONS needs to knwo if the EndPoitn set contains GSP_R or GSP_RW.
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java:512]: s/knwo/know and s/EndPoitn/Endpoint or EndPoint

**Review by kinow — APPROVED** [review]

File: jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java (hunk 3 above)
Comment (afs) [reply jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java:512]: Removed comment (it's not right anymore).
